controller/node_controller.go

```diff
@@ -1018,3 +1001,215 @@ func BackingImageDiskFileCleanup(node *longhorn.Node, bi *longhorn.BackingImage,
 		delete(bi.Spec.Disks, diskUUID)
 	}
 }
+
+func (nc *NodeController) checkMonitor(node *longhorn.Node) (monitor.Monitor, error) {
+	if node == nil {
+		return nil, nil
+	}
+
+	if nc.monitor != nil {
+		return nc.monitor, nil
+	}
+
+	monitor, err := monitor.NewNodeMonitor(nc.logger, nc.ds, node, nc.enqueueNodeForMonitor)
+	if err != nil {
+		return nil, err
+	}
+
+	nc.monitor = monitor
+
+	return monitor, nil
+}
+
+func (nc *NodeController) enqueueNodeForMonitor(key string) {
+	nc.queue.Add(key)
+}
+
+func (nc *NodeController) syncOrphans(node *longhorn.Node, collectedDataInfo map[string]*monitor.CollectedDiskInfo) error {
+	for diskName, diskInfo := range collectedDataInfo {
+		if err := nc.createOrphans(node, diskName, diskInfo); err != nil {
+			return errors.Wrapf(err, "failed to create orphans for disk %v", diskName)
+		}
+		if err := nc.deleteOrphans(node, diskName, diskInfo); err != nil {
+			return errors.Wrapf(err, "failed to delete orphans for disk %v", diskName)
+		}
+	}
+
+	return nil
+}
+
+func (nc *NodeController) deleteOrphans(node *longhorn.Node, diskName string, diskInfo *monitor.CollectedDiskInfo) error {
+	autoDeletionEnabled, err := nc.ds.GetSettingAsBool(types.SettingNameOrphanAutoDeletion)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get %v setting", types.SettingNameOrphanAutoDeletion)
+	}
+
+	orphans, err := nc.ds.ListOrphansRO()
+	if err != nil {
+		return errors.Wrap(err, "unable to list orphans")
+	}
+
+	for _, orphan := range orphans {
+		if orphan.Status.OwnerID != nc.controllerID {
+			continue
+		}
+
+		// If the on-disk data is missing, e.g. deleted by users by manual, we need to delete the orphaned CR.
+		// Orphan controller won't stat/check the on-disk data and isn't aware the of the missing replica directory,
+		// so we check and handle it here.
+		if orphan.Spec.Type == longhorn.OrphanTypeReplica {
+			onDiskDataExist, err := nc.isOnDiskDataExist(orphan, diskInfo.OrphanedReplicaDirectoryNames)
```

Comment:
I think this is redundant:
    1. If the dir suddenly disappears, the orphan controller will delete the CR only.
    2. The CR creation is based on `diskInfo.OrphanedReplicaDirectoryNames`. We don't need to verify it back and force.

Reply:
orphan controller does not track the existence of the orphaned directory, so the step is to check if the orphan CR's dir is still in the latest `diskInfo.OrphanedReplicaDirectoryNames`. If not, the orphan CR needs to be deleted.

Reply:
OK, got what you mean. It's like the alignment between the CRs and the collected dir list. Can you add a log before this deletion/alignment?

Reply:
okay